### PR TITLE
Uf 11062

### PR DESCRIPTION
**Pull request by AndreSoderlund**
## Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Removed feature
- [ ] Code style update (formatting etc.)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Content/Data

## Does this PR introduce a breaking change?
- [ ] Yes (I have stepped the version number accordingly)
- [x] No
      
## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have added/updated tests to cover my changes (if applicable).

### CODE REVIEW

File: src/main/java/se/sundsvall/checklist/service/EmployeeChecklistService.java [hunk 1]
```diff
@@ -357,4 +362,13 @@ private EmployeeChecklistResponse buildNoMatchResponse() {
 			.build();
 	}
 
+	public OngoingEmployeeChecklists getOngoingEmployeeChecklists(final String municipalityId, final ParameterPagingBase pagingBase) {
+		var pageable = PageRequest.of(pagingBase.getPage() - 1, pagingBase.getLimit(), pagingBase.sort());
+		var ongoingEmployeeChecklists = employeeChecklistIntegration.fetchAllOngoingEmployeeChecklists(municipalityId, pageable);
+
+		return OngoingEmployeeChecklists.builder()
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/checklist/service/EmployeeChecklistService.java:369]: Why not something like this

```java 
return employeeChecklistIntegration
.fetchAllOngoingEmployeeChecklists(municipalityId, pageable)
.map(result -> EmployeeMapper.toOngoingEmployeeChecklist(result))
.toList();
```
Then you wont get a NPE when you call `ongoingEmployeeChecklists.getContent()` if `fetchAllOngoingEmployeeChecklists` decides to return null :)

File: src/main/java/se/sundsvall/checklist/integration/db/EmployeeChecklistIntegration.java [hunk 2]
```diff
@@ -355,4 +358,16 @@ private Optional<ChecklistEntity> retrieveActiveChecklist(String municipalityId,
...
+	/**
+	 * Fetches all ongoing employee checklists for a specific municipality. An ongoing employee checklist is an employee
+	 * checklist that started before today and ends after today.
+	 *
+	 * @param  municipalityId the id of the municipality
+	 * @param  pageable       the page request
+	 * @return                a page of ongoing employee checklists
+	 */
+	public Page<EmployeeChecklistEntity> fetchAllOngoingEmployeeChecklists(final String municipalityId, final PageRequest pageable) {
+		return employeeChecklistRepository.findAllByChecklistsMunicipalityIdAndStartDateIsBeforeAndEndDateIsAfter(municipalityId, LocalDate.now().minusDays(1), LocalDate.now(), pageable);
```
Comment (josc04) [inline src/main/java/se/sundsvall/checklist/integration/db/EmployeeChecklistIntegration.java:371]: If you send in today minus 1 day the startdate will be interpreted as "checklist has started before yesterday" (StartDateIsBefore"TodayMinusOneDay") and not "started before today" - remove subtraction of 1 day to include all that started yesterday too?

File: src/main/java/se/sundsvall/checklist/api/model/ParameterPagingBase.java
Comment (josc04) [inline src/main/java/se/sundsvall/checklist/api/model/ParameterPagingBase.java:1]: Is this class needed. Can't you just use the AbstractParameterPagingAndSortingBase directly?

File: src/main/java/se/sundsvall/checklist/service/mapper/EmployeeChecklistMapper.java [hunk 3]
```diff
@@ -111,6 +112,25 @@ public static CustomTaskEntity updateCustomTaskEntity(CustomTaskEntity entity, C
...
 
+	public static List<OngoingEmployeeChecklist> toOngoingEmployeeChecklists(final List<EmployeeChecklistEntity> employeeChecklistEntities) {
+		return ofNullable(employeeChecklistEntities).orElse(emptyList()).stream()
+			.map(EmployeeChecklistMapper::toOngoingEmployeeChecklist)
+			.toList();
+	}
+
+	public static OngoingEmployeeChecklist toOngoingEmployeeChecklist(final EmployeeChecklistEntity employeeChecklistEntity) {
+		return ofNullable(employeeChecklistEntity).map(entity -> OngoingEmployeeChecklist.builder()
+			.withEmployeeName(entity.getEmployee().getFirstName() + " " + entity.getEmployee().getLastName())
```
Comment (josc04) [inline src/main/java/se/sundsvall/checklist/service/mapper/EmployeeChecklistMapper.java:123]: This makes the API somewhat hard to understand. The service returns a field named employeeName which is a combination of first and last name (same with manager name). But - to sort result by these two fields, the client needs to send in a combination of employee_firstName and employee_lastName (same procedure for manager).

Is it possible to use Spring database projection to combine first and last name of employee and manager (and possibly also use the projection to sort on the combined fields)?

Comment (AndreSoderlund) [reply src/main/java/se/sundsvall/checklist/service/mapper/EmployeeChecklistMapper.java:123]: I'm not sure that is possible, how would I include these projections in the pageable? I don't want to fetch all entities and manually create the paginated response.

Alternatively, we could create a database view with each ongoing employee checklist already mapped to the desired model. In that case, we could simplify the API and sort/page directly on that view.

File: src/main/java/se/sundsvall/checklist/integration/db/EmployeeChecklistIntegration.java [hunk 4]
```diff
@@ -355,4 +359,17 @@ private Optional<ChecklistEntity> retrieveActiveChecklist(String municipalityId,
...
+	 * Fetches all ongoing employee checklists for a specific municipality. An ongoing employee checklist is an employee
+	 * checklist that started before today and ends after today.
+	 *
+	 * @param  municipalityId the id of the municipality
+	 * @param  pageable       the page request
+	 * @return                a page of ongoing employee checklists
+	 */
+	public Page<OngoingEmployeeChecklistProjection> fetchAllOngoingEmployeeChecklists(final String municipalityId, final PageRequest pageable) {
+		var today = LocalDate.now();
+		return employeeChecklistRepository.findAllByChecklistsMunicipalityIdAndStartDateIsBeforeAndEndDateIsAfter(municipalityId, today, today, pageable);
```
Comment (josc04) [inline src/main/java/se/sundsvall/checklist/integration/db/EmployeeChecklistIntegration.java:373]: According to story it should be possible to filter the result should by name of new employee like %<value>%. The filter part seems to be missing?

**Review by josc04 — APPROVED** [review]

**Pull request merged**